LSPS3/README.md

```diff
@@ -0,0 +1,317 @@
+> client, the LSP does not have to remember what it offered via this
+> API, reducing state storage, especially if the client then decides
+> that the `opening_fee` is too high and ultimately decides not to
+> continue this flow.
+> By using a cryptographically-verifiable method, the LSP can ensure
+> that it will only honour `opening_fee`s it actually promised.
+> A size limit is imposed so that LSPs cannot burden clients with
+> unreasonable storage requirements.
+
+`min_payment_size` and `max_payment_size` are the limits of the
+payment size, inclusive.
+These are in millisatoshis.
+The payment size is the amount that the client will receive on the
+JIT Channel.
+
+The client MUST ignore any additional fields it does not recognize.
+
+> **Rationale** This allows future versions of this spec to add new
+> fields to this response, without causing problems for clients that
+> only support the current version.
+
+### 2.  Request JIT Channel
+
+The client constructs a request body for a `POST /lsp/jitchannel`
+request, depending on their need.
+
+Example `POST /lsp/jitchannel` request:
+
+```JSON
+{
+    "version": 1,
+    "opening_fee": "546000",
+    "opening_fee_valid_until": "2023-02-23T08:47:30.511Z",
+    "opening_fee_promise": "abcdefghijklmnopqrstuvwxyz",
+    "payment_size": "42000",
+    "client_node_id": "03d4e028a0d4a90868ec202ab684fb0085779defea9ca7553e06146557631eec20"
+}
+```
+
+`version` is the version of this spec that will be used for this
+interaction.
+
+`opening_fee`, `opening_fee_valid_until`, and `opening_fee_promise`
+are the parameters acquired from the previous step.
+Clients MUST copy them verbatim from some previous
+`GET /lsp/jitchannel` response.
+LSPs SHOULD check that the `opening_fee_promise` does in fact
+prove that it previously promised the specified `opening_fee` and
+`opening_fee_valid_until`.
+LSPs SHOULD check that the `opening_fee_valid_until` is not a
+past datetime.
+
+`payment_size` is the amount that the client wants to receive.
+
+The client MUST ensure that `payment_size` is within the previous
+`min_payment_size` and `max_payment_size` parameters from the LSP.
+The LSP MUST validate that the `payment_size` is within the previous
+`min_payment_size` and `max_payment_size` parameters from the LSP.
+
+The client MAY use a "subtractive" computation to get the
+`payment_size`.
+For example, the client may want to charge some payer an exact
+amount of 1,000,000 millisatoshis, but the LSP indicated an
+`opening_fee` of 546,000 millisatoshis.
+In that case, the client should specify 454,000 millisatoshis for
+`payment_size`, so that the invoice charges 1,000,000 millisatoshis
+total.
+
+`client_node_id` is the node ID of the client.
+
+The LSP can then respond with a response object or an error
+object.
+The client can differentiate between these two objects by checking
+for an `error` field, which exists and is the Boolean value
+`true` if the object is an error object.
```

Comment:
Can we add HTTP code here?
Proposal:
* `200` for success.
* `400` for request validation errors aka client error.

I guess we need to specify error more in general.

Reply:
I see your point. We are already including HTTP method (POST, GET) and paths (/lsp/jitchannels) though. What do you think of specifying a HTTP spec here with the requirement of it being easily portable to gRPC and onion?

For example, the `error` field here let's us use the spec in all protocols. HTTP codes would make it protocol specific. All protocol specific requirements MUST be removable without it impacting the function of the spec.

Reply:
I think that the initial version of the specs has to contain gRPC also. btw, gRPC has also a concept of status codes (see https://grpc.github.io/grpc/core/md_doc_statuscodes.html) but I agree that if we use HTTP status code or any other protocol specific feature we need to check that the semantic can be naturally translated to the other protocols.

Reply:
I assume it would be good to also have a gRPC spec for LSPS1?

Reply:
Yes, if we want to be consistent.

Reply:
LSPS0: https://github.com/BitcoinAndLightningLayerSpecs/lsp/pull/23

Reply:
LSPS0 merged.